frontend/src/components/dashboard/Analysis.tsx

```diff
@@ -119,6 +121,165 @@ const onLeaveLabel = (context: EventContext) => {
+              <div>
+                <span>{intl.formatMessage(messages.numberOfStudents)}</span>
+                <span>{group.analyses.length}</span>
+              </div>
+              <div>
+                <ul>
+                  {group.analyses.map((analysis) => (
+                    <li key={analysis.id}>
+                      <StudentNameWrapper
+                        onClick={() => onSelectSolution(group.key, analysis)}
+                        // add id for debugging purposes
+                        id={
+                          "analysis-" +
+                          analysis.id +
+                          "-solution-" +
+                          analysis.solutionId
+                        }
+                      >
+                        <StudentName
+                          pseudonym={analysis.studentPseudonym}
+                          keyPairId={analysis.studentKeyPairId}
+                          showActualName={false}
+                        ></StudentName>
+                      </StudentNameWrapper>
+                    </li>
+                  ))}
+                </ul>
+              </div>
+            </div>
+          ))}
+        </div>
+      ))}
+    </>
+  );
+};
+
+type AggregateDataPoint = {
+  x: number;
+  y: number;
+  dataPoints: CategorizedDataPoint[];
+};
+
+function aggregateDataPointsByCoordinate(
+  categorizedDataPoints: CategorizedDataPoint[],
+): AggregateDataPoint[] {
+  // first turn the data points into a map of maps (X,Y) -> data points
+  const byCoordinates = categorizedDataPoints.reduce((acc, dataPoint) => {
+    const byX = acc.get(dataPoint.x);
+    if (byX) {
+      const byY = byX.get(dataPoint.y);
+
+      if (byY) {
+        byY.push(dataPoint);
+      } else {
+        byX.set(dataPoint.y, [dataPoint]);
+      }
+    } else {
+      acc.set(dataPoint.x, new Map([[dataPoint.y, [dataPoint]]]));
+    }
+
+    return acc;
+  }, new Map<number, Map<number, CategorizedDataPoint[]>>());
+
+  // then flatMap the map of maps to a list of objects
+  const mergedDataPoints = byCoordinates
+    .entries()
+    .flatMap(([x, byY]) =>
+      byY.entries().map(([y, dataPoints]) => ({
+        x,
+        y,
+        dataPoints,
+      })),
+    )
+    .toArray();
+
+  return mergedDataPoints;
+}
+
+function buildGroupsByKey(
+  dataPoints: CategorizedDataPoint[],
+): { key: string; name: string; analyses: CurrentAnalysis[] }[] {
+  return Object.entries(
+    dataPoints.reduce(
+      (byGroupKey, dataPoint) => {
+        if (dataPoint.groupKey in byGroupKey) {
+          byGroupKey[dataPoint.groupKey].push(dataPoint);
+        } else {
+          byGroupKey[dataPoint.groupKey] = [dataPoint];
+        }
+
+        return byGroupKey;
+      },
+      {} as { [groupKey: string]: CategorizedDataPoint[] },
+    ),
+  ).map(([key, groupDataPoints]) => ({
+    key,
+    name: groupDataPoints[0].groupName,
+    analyses: groupDataPoints.flatMap((d) => d.analyses ?? []),
+  }));
+}
```

Comment:
All this code was extracted as-is from other places.
It's a first step in a broader refactoring.